### PR TITLE
ddl: add format error for incorrect dict syntax in the placement rule

### LINKED ISSUE #30454: Can we improve the error message of incorrect dict syntax?
This is currently documented as a YAML parser issue:
```sql
tidb> CREATE PLACEMENT policy p4 LEADER_CONSTRAINTS="[+region=us-east-1]" FOLLOWER_CONSTRAINTS="{+region=us-east-2:2}" FOLLOWERS=2;
..
ERROR 1105 (HY000): label constraints in map syntax have invalid replicas: count of labels '+region=us-east-2:2' should be positive, but got 0: invalid FollowerConstraints

```

It needs to be specified instead as:
```sql
tidb> CREATE PLACEMENT policy p4 LEADER_CONSTRAINTS="[+region=us-east-1]" FOLLOWER_CONSTRAINTS="{+region=us-east-2: 2}" FOLLOWERS=2;
Query OK, 0 rows affected (0.10 sec)
```

I think this is going to be a common issue. If we can't fix it, can we improve the error message?

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #30454

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/placement/rule_test.go [hunk 1]
```diff
@@ -50,15 +48,32 @@ func matchRules(t1, t2 []*Rule, prefix string, c *C) {
...
+func TestNewRuleAndNewRules(t *testing.T) {
 	type TestCase struct {
 		name     string
 		input    string
 		replicas uint64
 		output   []*Rule
 		err      error
 	}
-	tests := []TestCase{}
+	var tests []TestCase
```
Comment (hawkingrei) [inline ddl/placement/rule_test.go:76]: Goland raises a warning here.

**Comment by hawkingrei** [comment]
@lcwangchao @morgo PTAL

File: ddl/placement/rule_test.go [hunk 2]
```diff
@@ -50,15 +48,32 @@ func matchRules(t1, t2 []*Rule, prefix string, c *C) {
 	}
 }
 
-func (t *testRuleSuite) TestNewRuleAndNewRules(c *C) {
+func matchRulesT(t1, t2 []*Rule, prefix string, t *testing.T) {
+	require.Equal(t, len(t2), len(t1), prefix)
+	for i := range t1 {
+		found := false
+		for j := range t2 {
+			ok, _ := DeepEquals.Check([]interface{}{t2[j], t1[i]}, []string{})
```
Comment (AilinKid) [inline ddl/placement/rule_test.go:56]: DeepEquals depend on `check`, could we remove it here too

File: ddl/placement/rule_test.go [hunk 3]
```diff
@@ -137,7 +152,7 @@ func (t *testRuleSuite) TestNewRuleAndNewRules(c *C) {
 		name:     "zero count in object constraints",
 		input:    `{"+zone=sh,-zone=bj":0, "+zone=sh": 1}`,
 		replicas: 3,
-		err:      ErrInvalidConstraintsMapcnt,
+		err:      ErrMappingWrongSeparator,
```
Comment (xhebox) [inline ddl/placement/rule_test.go:155]: This is not a case of wrong separator, since `"..":4` is valid.

File: ddl/placement/rule.go [hunk 4]
```diff
@@ -62,6 +62,23 @@ func NewRule(role PeerRoleType, replicas uint64, cnst Constraints) *Rule {
...
+				return ErrMappingWrongSeparator
+			}
+			findSeparator = false
+			if str[i] == ':' {
+				findSeparator = true
+			}
+		}
+		return nil
+	}
+	return ErrMappingNoColonFound
```
Comment (xhebox) [inline ddl/placement/rule.go:79]: I recommended to use a pre compiled regexp like `[^"':]+:\d` to match.

File: ddl/placement/rule.go [hunk 5]
```diff
@@ -87,6 +104,9 @@ func NewRules(role PeerRoleType, replicas uint64, cnstr string) ([]*Rule, error)
 		ruleCnt := 0
 		for labels, cnt := range constraints2 {
 			if cnt <= 0 {
+				if err := getYamlMapFormatError(string(cnstbytes)); err != nil {
+					return rules, err
```
Comment (xhebox) [inline ddl/placement/rule.go:108]: I think it is better to do `return rules, fmt.Errorf("%w: %s", ErrInvalidConstraintsMapcnt, err)`, or even just `return rules, fmt.Errorf("%w:  wrong separator ...", ErrInvalidConstraintsMapcnt)`. We don't need to add a new error for every kind.

**Review by xhebox — COMMENTED** [review]
BTW, you can parallel tests in this suite. Most tests in placement packages are parallelable.

File: ddl/placement/rule.go [hunk 6]
```diff
@@ -62,6 +62,23 @@ func NewRule(role PeerRoleType, replicas uint64, cnst Constraints) *Rule {
 	}
 }
 
+func getYamlMapFormatError(str string) error {
+	if strings.Contains(str, ":") {
```
Comment (lcwangchao) [inline ddl/placement/rule.go:66]: May be it's more brief

```
	if !strings.Contains(str, ":") {
		return ErrMappingNoColonFound
	}
	
	for i := 0; i < len(str) - 1; i = i + 1 {
		if str[i] == ':' && str[i+1] != ' ' {
			return ErrMappingWrongSeparator
		}
	}
	
	return nil
```

File: ddl/placement/rule.go (hunk 4 above)
Comment (hawkingrei) [reply ddl/placement/rule.go:79]: I think you are right.

**Comment by hawkingrei** [comment]
> BTW, you can parallel tests in this suite. Most tests in placement packages are parallelable.

We are adding the linter to avoid ```t.Parallel()``` and we will use the @tiancaiamao 's tool to make the test suit parallel.

File: ddl/placement/rule_test.go (hunk 2 above)
Comment (hawkingrei) [reply ddl/placement/rule_test.go]: Done

File: ddl/placement/rule_test.go (hunk 3 above)
Comment (hawkingrei) [reply ddl/placement/rule_test.go]: Done

File: ddl/placement/rule.go (hunk 4 above)
Comment (hawkingrei) [reply ddl/placement/rule.go]: Done

File: ddl/placement/rule.go (hunk 5 above)
Comment (hawkingrei) [reply ddl/placement/rule.go]: Done

Comment (hawkingrei) [reply ddl/placement/rule.go]: ErrInvalidConstraintsMapcnt is too wordy for users to provide solutions. So "wrong separator ..." is better.

File: ddl/placement/rule.go (hunk 6 above)
Comment (hawkingrei) [reply ddl/placement/rule.go]: Done

File: ddl/placement/rule.go (hunk 5 above)
Comment (hawkingrei) [reply ddl/placement/rule.go:104]: Done

**Review by xhebox — APPROVED** [review]

**Review by lcwangchao — APPROVED** [review]
